### DISCUSSION TITLE
Getting ETH + ZIL dual mining working on RBM without Ezil, with working examples

### DISCUSSION THREAD

**Discussion by coolbreeze21** [RainbowMiner/RainbowMiner#1941] [Show and tell]
TL;DR version: It is possible to run ETH + ZIL dual mining under RBM using Rustpool (and probably Shardpool) as a proxy and 2MinersAE (and probably others) as the ETH pool.  Details below.

Before I discovered RBM, I used to dual mine ETH + ZIL on my GPUs.  It was pretty clear to me that allowing the GPU to switch from ETH to ZIL for roughly one minute every hour had a net improvement in profitability.  Not the “up to 15%” that Ezil claims, but a few percent.

I tried getting this working with RBM, but ran into several problems as listed below.  I wanted to do it using 2MinersAE as the ETH pool, because of ETH gas prices on payouts having gone to the moon (and pathetic alternative Polygon having been brought to its knees by one laughably primitive play-to-earn Ponzi game, SFF).

1.	The Ezil.me documentation makes it pretty clear that they do not allow you to mine ZIL there and mine ETH on another pool.  All my attempts to do so were failures.  I am pretty sure that the RBM Readme.md instructions about being able to do this using a dummy ETH address are incorrect or no longer correct.  Mining both ETH + ZIL at Ezil.me is not attractive to me for several reasons, including high gas prices on payouts.

So, I needed to pick an alternative ZIL pool that allows mining ETH to another pool.  The only alternatives with any decent hashrate are shardpool.io and zil.rustpool.xyz.  Shardpool has a 1% pool fee.  Rustpool has no pool fee, a low 10 ZIL minimum payout, and I had used them before, so I focused on Rustpool, although I did get Shardpool working as well.

2.	The RBM Readme.md instructions say to use Nanominer.  But although Nanominer is available on my rigs for multiple algos, Ethash isn’t one of them (the same strange situation is true of TeamBlack).  I don’t know what causes this or how to fix it, so this is a nonstarter for me.

3.	A closer look at the RBM json settings reveals that, in addition to Nanominer, Lolminer and Trex also have “enable dual mining” settings.  Since both are available on my rigs for Ethash, I started with Trex, which I had used successfully before for dual mining.  RBM can launch Trex, connect with 2Miners and load both DAGs successfully, but it fails to connect with Rustpool.  The error message suggests to me there is a syntax error in RBM’s string calling for Trex to connect to Rustpool, but that is beyond my ability to diagnose or fix, so this attempt failed at this point.  

4.	RBM can launch Lolminer and connect to both 2MinersAE and Rustpool without throwing any errors, but I never saw it mine any ZIL, even after several ZIL rounds.  Even worse, after returning from a ZIL round, Lolminer would sometimes stop mining ETH shares that were credited on the pool side, although Lolminer and RBM thought everything was fine.  Not good, to put it mildly.

So far, everything above is a boatload of fail.  Now for some success:

I decided to go back to what worked for me before I used RBM, which was using the ZIL pool as a proxy to the ETH pool.  This requires setting up either Rustpool or Shardpool as a custom pool.  Below are my userpools.config.txt settings for Rustpool, using either PhoenixMiner or Trex as the miner (Rustpool will not work with Gminer according to my testing and their Telegram support channel).  Rustpool has 2 servers in the US, 3 in the EU and 3 in Asia – use your region where appropriate below.

"Enable": "1",
"Name": "Rustpool",
"PoolFee": "0",
"Algorithm": "Ethash",
"CoinName": "",
"CoinSymbol": "ETH",
"Currency": "BTC",
"Protocol": "stratum+tcp",
"Host": "us2-zil.rustpool.xyz",
"Port": "8008",
"User": "$Wallet", [or BTC or NANO wallet address, or ETH address if you like paying ridiculous gas fees]
"Pass": "[ZIL wallet]@us-eth.2miners.com:2020@4G", [for Phoenix, use 8G if rig higher than 180Mh/s]
"Pass": "[ZIL wallet]@us-eth.2miners.com:2020@4G --extra-dag-epoch 0", [for Trex, 8G as above]
"Region": "US",
"SSL": "0",
"EthMode": "ethproxy"

Once I saved the above in userpools.config.txt and allowed RBM to add the Rustpool entry to pools.confiIg.txt, I checked or made changes to the additional settings below in the Rustpool entry of pools.confiIg.txt.  This required editing the pools.config.txt directly, since RBM didn’t make them available via “Setup Pools”.

"BTC": "$Wallet", [may have been unnecessary]
"Worker": "[your worker name]",
"CoinSymbol": "ETH", [may have been unnecessary, but I didn’t want to inadvertently mine another coin.  May also be necessary to avoid switching DAGs when Phoenix mines dev fee ]
"MinerName": "[either Phoenix or Trex]", [needed to force the use of the miner/launch string selected in “Pass” above and no other]

All other pools.config.txt Rustpool settings were left untouched.

Finally, I needed to disable all other pools that were enabled to mine ETH.  Because Rustpool is a custom pool, it is getting its profitability stats from WhatToMine and WTM generally returns lower profitability numbers for ETH than the alternatives.  In my case, this only meant unchecking 2MinersAE in Setup and excluding ETH as a coin in in MoneroOcean (MO is a poor choice anyway, since they almost never find an ETH block).

It works!  As long as RBM sees ETH (according to WTM) as the most profitable coin, it will mine it on 2MinersAE via Rustpool as a proxy and automatically switch to mining ZIL for about a minute every hour.  It will also switch back successfully after the ZIL minute is over and continue mining ETH (unlike the Lolminer example above).  RBM doesn’t track the ZIL, but you can see it adding up on the Rustpool website.  Rustpool pays out once a day whenever your balance exceeds 10 ZIL.  This has been tested over several days, all the way through a successful payout.

I am pretty sure the above can be adapted to Shardpool as the ZIL pool.  It did work using bat files outside RBM.  One advantage is that Gminer, the fastest ETH miner on my rigs, works on Shardpool.  One disadvantage is that Shardpool has an added 1% pool fee.

It may seem like a lot of fuss to squeeze out an extra few percent of profitability but, if you are already using RBM for algo-switching/pool hopping, you are probably interested in trying to squeeze out as much profitability as possible.

So, if anyone else is interested in mining ETH + ZIL, there are a few examples of how to do it.  If you have other or better ways of doing it, please reply and share your ideas.  This was the best I could get working at this point.

Many thanks to the developer of RainbowMiner!

**Comment by RainbowMiner** [comment]
Thank you very much for this post! It helps a lot. In fact, there is a lot of information between the lines that point to some smaller flaws in the core code, that are for me to fix.

> It may seem like a lot of fuss to squeeze out an extra few percent of profitability but, if you are already using RBM for algo-switching/pool hopping, you are probably interested in trying to squeeze out as much profitability as possible.

Every extra penny is worth the work, in my opinion.
